docs/design/gluster-block-provisioning.md

```diff
@@ -0,0 +1,306 @@
+
+The API is a variation of the ```volume``` API and looks like this.
+
+The supported functions are:
+
+* ```BlockVolumeCreate```
+* ```BlockVolumeInfo```
+* ```BlockVolumeDelete```
+* ```BlockVolumeList```
+
+In the future, ```BlockVolumeExpand``` might get added.
+
+### Note About The State Of The Implementation
+
+As of 2017-05-19, a WIP API and overall skeleton implementation is available in
+<https://github.com/obnoxxx/heketi/tree/block-wip>.
+
+
+### Details about the requests
+
+#### BlockVolumeCreateRequest
+
+The block volume create request takes the size and a name
+and can optionally take a list of clusters and an hacount.
+
+```
+type BlockVolumeCreateRequest struct {
+        Size       int       `json:"size"`
+        Clusters   []string  `json:"clusters,omitempty"`
+        Name       string    `json:"name"`
+        Hacount    int       `json:"hacount,omitempty"`
+}
+```
+
+#### BlockVolume
+
+This is the basic info about a block volume.
+
+```
+BlockVolume struct {
+        Hosts     []string `json:"hosts"`
+        Hacount   int `json:"hacount"`
+        Iqn       string `json:"iqn"`
+        Lun       int `json:"lun"`
+}
+
+```
+
+#### BlockVolumeInfo
+
+This is returned for the blockvolume info request and
+upon successful creation.
+
+```
+type BlockVolumeInfo struct {
+        Size       int       `json:"size"`
+        Clusters   []string  `json:"clusters,omitempty"`
+        Name       string    `json:"name"`
+        Hacount    int       `json:"hacount,omitempty"`
+        Id         string    `json:"id"`
+        Size       int       `json:"size"`
+        BlockVolume struct {
+                Hosts     []string `json:"hosts"`
+                Hacount   int `json:"hacount"`
+                Iqn       string `json:"iqn"`
+                Lun       int `json:"lun"`
+        } `json:"blockvolume"`
+}
+
+```
+
+
+#### BlockVolumeListResponse
+
+The block volume list request just gets a list
+of block volume IDs as response.
+
+
+```
+type BlockVolumeListResponse struct {
+        BlockVolumes []string `json:"blockvolumes"`
+}
+
+```
+
+### Details About Heketi's Internal Logic
+
+Volumes used for gluster-block volumes should not be used
+for other purposes. Heketi needs to find such volumes.
+Hence heketi should internally flag these volumes with a
+label (```block```).
+
+By default, heketi would consider all available clusters
+when looking for space to create a new block-volume file.
+
+With the clusters request parameter, this search can be
+narrowed down to an explicit list of one or more clusters.
+With the help of the ```clusterid``` storage class option,
+this gives the kubernetes administrator a way to e.g. separate
+different storage qualities, or to reserve a cluster
+exclusively for block-volumes.
+
+#### Question
+
+Should Heketi create new file volumes for hosting block
+volume files (or in the future try to expand the currently
+used file volume, or is that solely up to the administrator?
```

Comment:
I feel that Heketi should create new file volumes if none of the existing volumes can satisfy the block request.